wgsl/index.bs

```diff
@@ -6983,11 +6981,283 @@ TODO: *Stub*: Expression evaluation
+This way, whenever a function is analyzed, the metadata for all of its callees has already been computed.
+There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+Note: another way of saying the same thing is that we do a topological sort of functions ordered by the "is a (possibly indirect) callee of" partial order, and analyze in that order.
+
+### Analyzing the uniformity requirements of a function ### {#uniformity-function}
+
+Each function is analyzed in two phases.
+
+The first phase walks over the syntax of the function, building a directed graph along the way based on the rules in the following subsections.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+<div class="note">Note: apart from two special nodes MustBeUniform and CannotBeUniform, all nodes can be understood as having one of the following meanings:
+        - The control flow at a specific program point must be uniform
+        - The value of a specific expression must be uniform
+        - A specific variable must always hold a uniform value
+
+        An edge can be understood as an implication from the statement corresponding to its source node to the statement corresponding to its target node.
+
+        To express that something must always be uniform (e.g. the control flow at the call site of a derivative), we simply add an edge from MustBeUniform to the corresponding node.
+        One way to understand this, is that MustBeUniform corresponds to the proposition True, so that MustBeUniform -> X is the same as saying that X is true.
+
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we simply add an edge from the corresponding node to CannotBeUniform.
+        One way to understand this, is that CannotBeUniform corresponds to the proposition False, so that X -> CannotBeUniform is the same as saying that X is false.
+
+        A consequence of this interpretation is that every node reachable from MustBeUniform corresponds to something which must be uniform for the program to be valid, and every node from which cannotBeUniform is reachable corresponds to something whose uniformity we cannot guarantee. It follows that we have a uniformity violation (and thus reject the program) if there is any path from MustBeUniform to CannotBeUniform.
+</div>
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some nodes called "MustBeUniform", "CannotBeUniform", "CF_start", "CF_return" and if the function is non-void a node called "Value_return"
+* Create one node for each parameter of the function which we'll call "arg_i"
+* Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next section.
+* Look at which nodes are reachable from "MustBeUniform"
```

Comment:
This took several passes to understand what it means. It might help to describe the data type that we remember a bit more like code, and provide signification to all of them. For example for each function remember (assuming `Uniformity` is `enum {None, Uniform}` but can grow more detailed to `Quad/Workgroup/SubgroupUniform` in the future):

 - A `callUniformityRequirement : Uniformity` that describes the requirement on this function's caller CF at the moment of call.
 - A `argumentUniformityRequirement: List<Uniformity>`, one for each argument that describes the requirement on this function's caller expression uniformity requirements on the moment of the call.
 - Both for `exitCF` and `returnValue`:
   - A `bestUniformityAtExit : Uniformity`
   - `argumentForExitUniformity : List<int>`
   - Note this will be used to compute the uniformity when returning to the caller with `worse([bestUniformityAtExit] + [uniformity of arg(i) for i in argumentForExitUniformity])`

This would help the readers (or at least me), build a mental model of how the analysis handles functions before getting into the details below because we have the semantic of what's happening already.

Reply:
Thanks for the suggestion, I've applied it (with some tweaks since we track a bit more information than that). Please have a look and tell me whether you find the result more readable. The names I've given the tags are a bit verbose, but hopefully give the intuition of why this or that specific piece of information matters.

Reply:
I think you forgot to push  the changes to Github :)

Reply:
I've started reviewing in this detail and had very similar confusion and feedback.

The "MustBeUniform" name is confusing, because it reads as two different things:
- a **requirement** that control flow be uniform, or
- a **conclusion** reached by the analysis, i.e. analysis has deduced that control flow is uniform at that point.

Given the rest of the text I think the correct interpretation is as a **requirement**.

Similarly MustBeUniformForControlFlowAfter is baffling and I don't have a great handle on what that really means.  My current working interpretation is "this node does not itself impose a uniformity requirement, but may propagate a requirement that comes from dynamically later in the control flow graph". And if that's the case then I would tentatively call it "PropagatesUniformityRequirement" or similar.

Reply:
Similarly, "CannotBeUniform" in my mind is "AssumedNonUniform".
That better captures the uncertainty / conservatism in the analysis.
There are two cases:
- the analysis *knows* the value is not uniform, e.g. local id in a compute shader
- the analysis cannot tell, and must assume non-uniformity, e.g. the value loaded from a shared buffer variable.
      - this latter case is marked as CannotBeUniform in the uniformity rules in the expressions, but it's confusing to me to name it that way.  It would be better as "AssumedNonUniform".


Then, for example, the easy rejection of a program is if "AssumedNonUniform" is reachable from "RequiredUniform", and that makes intuitive sense just from its wording.

Reply:
That is correct, "MustBeUniform" means "required to be uniform for the program to be accepted" and "CannotBeUniform" means "we cannot prove that this is uniform". "MustBeUniformForControlFlowAfter" means "required to be uniform if we need to prove that the control-flow after is uniform".

I apologize for any confusion, and am totally open to any renaming/rephrasing. I found it kind of tricky to make this clear, especially as it is completely clear to me after so much time reading/writing it.

Reply:
I expressed a similar concern in a comment, glad this is noticed. Let's have good names here, please.
"CannotBeUniform" -> "MayBeNonUniform" or "AssumedNonUniform"
"MustBeUniformForControlFlowAfter" -> "PropagatesUniformity" or something along the lines (edit: I see @Kangz suggested a very similar thing earlier)